resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation,
+modified to account for explicit nullability and the accompanying type system
+changes (including the legacy types).  In the interest of backwards
+compatibility, it does not try to fix the various issues with the existing
+algorithm.  
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
```

Comment:
> That document does not mention multiple modifiers, like `T??` or `T?*`, but they were hinted at yesterday, so I'm not sure about their status.

These are handled inductively - for a type `T?` `T` itself may always be of the form `S?` for some `S`.

> Another thing is that basically all rulesets below are defined in such a way that they _must_ assume prioritization: Every non-first rule starts with 'otherwise'. I think this should be mentioned.

Fixed by merge from the non NNBD version, let me know if you think it needs more commentary.